client/utils/src/pubsub.rs

```diff
@@ -0,0 +1,262 @@
+//! - `Dispatch<M>`;
+//! - `Unsubscribe`.
+//!
+//! As a result of subscription `Hub::subscribe` method returns an instance of `Receiver<Message,
+//! Registry>`. That can be used as a `futures::Stream` to receive the messages.
+//! Upon drop the `Receiver<Message, Registry>` shall unregister itself from the `Hub`.
+
+use std::{
+	collections::HashMap,
+	ops::DerefMut,
+	pin::Pin,
+	sync::{Arc, Weak},
+	task::{Context, Poll},
+};
+
+use ::futures::stream::{FusedStream, Stream};
+use ::parking_lot::Mutex;
+
+use crate::mpsc::{TracingUnboundedReceiver, TracingUnboundedSender};
+
+// pub mod channels;
+
+/// The type to identify subscribers.
+pub type SubsID = crate::id_sequence::SeqID;
+
+/// Unsubscribe: unregisters a previously created subscription.
+pub trait Unsubscribe {
+	/// Remove all registrations of the subscriber with ID `subs_id`.
+	fn unsubscribe(&mut self, subs_id: &SubsID);
+}
+
+/// Subscribe using a key of type `K`
+pub trait Subscribe<K> {
+	/// Register subscriber with the ID `subs_id` as having interest to the key `K`.
+	fn subscribe(&mut self, subs_key: K, subs_id: SubsID);
+}
+
+/// Dispatch a message of type `M`.
+pub trait Dispatch<M> {
+	/// The type of the that shall be sent through the channel as a result of such dispatch.
+	type Item;
+	/// The type returned by the `dispatch`-method.
+	type Ret;
+
+	/// Dispatch the message of type `M`.
+	///
+	/// The implementation is given an instance of `M` and is supposed to invoke `dispatch` for
+	/// each matching subscriber, with an argument of type `Self::Item` matching that subscriber.
+	///
+	/// Note that this does not have to be of the same type with the item that will be sent through
+	/// to the subscribers. The subscribers will receive a message of type `Self::Item`.
+	fn dispatch<F>(&mut self, message: M, dispatch: F) -> Self::Ret
+	where
+		F: FnMut(&SubsID, Self::Item);
+}
+
+/// A subscription hub.
+///
+/// Does the subscription and dispatch.
+/// The exact subscription and routing behaviour is to be implemented by the Registry (of type `R`).
+/// The Hub under the hood uses the channel defined in `crate::mpsc` module.
+#[derive(Debug)]
+pub struct Hub<M, R> {
+	tracing_key: &'static str,
+	shared: Arc<Mutex<Shared<R, TracingUnboundedSender<M>>>>,
+}
+
+/// The receiving side of the subscription.
+///
+/// The messages are delivered as items of a `futures::Stream`.
+/// Upon drop this receiver unsubscribes itself from the `Hub`.
+#[derive(Debug)]
+pub struct Receiver<M, R>
+where
+	R: Unsubscribe,
+{
+	// NB: this field should be defined before `rx`.
+	// (The fields of a struct are dropped in declaration order.)[https://doc.rust-lang.org/reference/destructors.html]
+	_unsubs_guard: UnsubscribeGuard<R, TracingUnboundedSender<M>>,
+
+	// NB: this field should be defined after `_unsubs_guard`.
+	rx: TracingUnboundedReceiver<M>,
+}
+
+#[derive(Debug)]
+struct UnsubscribeGuard<R, Tx>
+where
+	R: Unsubscribe,
+{
+	shared: Weak<Mutex<Shared<R, Tx>>>,
```

Comment:
What is the gain in resorting to dynamic dispatch? Just to erase the `R` type out of the receiver part?

I am not convinced that would be better:
- would need to introduce some `UnregisterViaImmutableSelfRef` trait;
- define that trait over `Mutex<Shared<R, ...>> where R: Unsubscribe`;

Having both the existing `Unsubscribe` trait and the `UnregisterViaImmutableSelfRef` trait declared wouldn't be that clear to a reader (in my opinion).

Having a trait `Unsubscribe` to be implemented by the one who implements another pub/sub, while having a `UnregisterViaImmutableSelfRef` declared just as a technicality to perform that dyn-call — it would smell from my point of view.

Reply:
> What is the gain in resorting to dynamic dispatch? Just to erase the `R` type out of the receiver part?

Yes.

Because the `Receiver` is a public API, meant to be used outside of this module in many places, and generics in Rust are infectious, so instead of simply writing e.g. `fn foobar<M>(rx: Receiver<M>)` you'll have to write `fn foobar<M, R>(rx: Receiver<M, R>) where R: Unsubscribe` (or have to put the concrete type in) when there's literally no practical benefit in having that `R` there.

I don't think making the public API (mean to be used multiple times in multiple places) more ugly and more inconvenient to use is not a good tradeoff in exchange for just not having to define a single extra *internal* trait in one place.

Reply:
The current usage (and such before this PR) of the notifications introduced a type def for its particular case.

Thus I'd insist, the example of `fn foobar<M, R>(rx: Receiver<M, R>)` is a bit far-fetched. 
Realistically this function is more likely defined as `fn foobar(rx: GrandpaJustificationStream)`.

So the details of "having some registries inside" never leave the region of code where it is actually decided, that some registries are used (e.g. `sc-finality-grandpa::notifications`).


Generics in this particular case do not have to be infectious: one could just define the function as `fn foobar<S>(events_stream) where S: Stream<Item = ANotification>`; in your example — you chose to bring the whole internal machinery into the function's signature, although you did not have to.

Reply:
I disagree. (: 

But very well, I'm not going to force it.

Reply:
Thanks for your input though. :)